project/JarJar.scala

```diff
@@ -0,0 +1,83 @@
+import com.tonicsystems.jarjar
+import com.tonicsystems.jarjar._
+import com.tonicsystems.jarjar.ext_util._
+import scala.collection.JavaConverters._
+import java.util.jar._
+import java.io._
+import sbt._
+
+object JarJar {
+  sealed abstract class JarJarConfig {
+    def toPatternElement: PatternElement
+  }
+  object JarJarConfig {
+    case class Rule(pattern: String, result: String) extends JarJarConfig {
+      def toPatternElement: PatternElement = {
+        val rule = new jarjar.Rule
+        rule.setPattern(pattern)
+        rule.setResult(result)
+        rule
+      }
+    }
+    case class Keep(pattern: String) extends JarJarConfig {
+       def toPatternElement: PatternElement = {
+          val keep = new jarjar.Keep
+          keep.setPattern(pattern)
+          keep
+       }
+    }
+  }
+
+  sealed abstract class Entry {
+    def name: String
+    def time: Long
+    def data: Array[Byte]
+  }
+
+  case class JarEntryInput(jarFile: JarFile, entry: JarEntry) extends Entry {
+    def name = entry.getName
+    def time = entry.getTime
+    def data = sbt.IO.readBytes(jarFile.getInputStream(entry))
+  }
+  case class FileInput(base: File, file: File) extends Entry {
+    def name = file.relativeTo(base).get.getPath
+    def time = file.lastModified
+    def data = sbt.IO.readBytes(file)
+  }
+
+  private def newMainProcessor(patterns: java.util.List[PatternElement], verbose: Boolean, skipManifest: Boolean): JarProcessor = {
+    val cls = Class.forName("com.tonicsystems.jarjar.MainProcessor")
+    val constructor = cls.getConstructor(classOf[java.util.List[_]], java.lang.Boolean.TYPE, java.lang.Boolean.TYPE)
+    constructor.setAccessible(true)
+    constructor.newInstance(patterns, Boolean.box(verbose), Boolean.box(skipManifest)).asInstanceOf[JarProcessor]
+  }
+
+  def apply(in: Iterator[Entry], outdir: File,
+            config: Seq[JarJarConfig], verbose: Boolean = false): Seq[File] = {
+    val patterns = config.map(_.toPatternElement).asJava
+    val processor: JarProcessor = newMainProcessor(patterns, verbose, false)
+    def process(e: Entry): Option[File] = {
+      val struct = new EntryStruct()
+      struct.name = e.name
+      struct.time = e.time
+      struct.data = e.data
+      if (processor.process(struct)) {
+        if (struct.name.endsWith("/")) None
+        else {
+          val f = outdir / struct.name
+          try {
+            f.getParentFile.mkdirs()
+            sbt.IO.write(f, e.data)
```

Comment:
Here's the source of the last bug you mentioned: This should be `struct.data` instead of `e.data`